src/fuselage/CCPACSSkinSegment.cpp

```diff
@@ -0,0 +1,168 @@
+
+#include "CCPACSSkinSegment.h"
+
+#include <TopoDS_Face.hxx>
+#include <TopoDS_Edge.hxx>
+#include <BRep_Tool.hxx>
+#include <BRepExtrema_ExtCC.hxx>
+#include <TopExp.hxx>
+#include <BRepTools.hxx>
+#include <Bnd_Box.hxx>
+#include <BRepBndLib.hxx>
+#include <TopTools_IndexedMapOfShape.hxx>
+#include <Geom_Surface.hxx>
+
+#include "generated/CPACSSkinSegments.h"
+#include "CCPACSFrame.h"
+#include "CCPACSFuselageStringer.h"
+#include "CTiglError.h"
+#include "CCPACSFuselage.h"
+#include "CTiglUIDManager.h"
+#include "CNamedShape.h"
+#include "tiglcommonfunctions.h"
+
+namespace tigl
+{
+CCPACSSkinSegment::CCPACSSkinSegment(CCPACSSkinSegments* parent, CTiglUIDManager* uidMgr)
+    : generated::CPACSSkinSegment(parent, uidMgr)
+{
+}
+
+void CCPACSSkinSegment::Invalidate()
+{
+    m_borderCache = boost::none;
+}
+
+bool CCPACSSkinSegment::Contains(const TopoDS_Face& face)
+{
+    double u_min = 0., u_max = 0., v_min = 0., v_max = 0.;
+    BRepTools::UVBounds(face, u_min, u_max, v_min, v_max);
+    Handle(Geom_Surface) testSurf = BRep_Tool::Surface(face);
+    gp_Pnt point                  = testSurf->Value(u_min + ((u_max - u_min) / 2), v_min + ((v_max - v_min) / 2));
+    return Contains(point);
+}
+
+bool CCPACSSkinSegment::Contains(const TopoDS_Edge& edge)
+{
+    double u_min = 0., u_max = 0.;
+    Handle(Geom_Curve) curve = BRep_Tool::Curve(edge, u_min, u_max);
+    gp_Pnt point             = curve->Value(u_min + ((u_max - u_min) / 2));
+    return Contains(point);
+}
+
+bool CCPACSSkinSegment::Contains(const gp_Pnt& point)
+{
+    if (!m_borderCache)
+        UpdateBorders();
+
+    const double angleLower = Radians(45.);
+    const double angleUpper = Radians(89.);
+
+    const BorderCache& c = m_borderCache.value();
+    if (gp_Ax1(c.sFrame_sStringer.Location(), gp_Vec(c.sFrame_sStringer.Location(), point)).Angle(c.sFrame_sStringer) >= angleLower) return false;
+    if (gp_Ax1(c.sFrame_eStringer.Location(), gp_Vec(c.sFrame_eStringer.Location(), point)).Angle(c.sFrame_eStringer) >= angleLower) return false;
+    if (gp_Ax1(c.eFrame_sStringer.Location(), gp_Vec(c.eFrame_sStringer.Location(), point)).Angle(c.eFrame_sStringer) >= angleLower) return false;
+    if (gp_Ax1(c.eFrame_eStringer.Location(), gp_Vec(c.eFrame_eStringer.Location(), point)).Angle(c.eFrame_eStringer) >= angleLower) return false;
+
+    const gp_Pnt p1 = (c.sFrame_sStringer.Location().XYZ() + c.eFrame_sStringer.Location().XYZ()) / 2;
+    const gp_Pnt p2 = (c.eFrame_sStringer.Location().XYZ() + c.eFrame_eStringer.Location().XYZ()) / 2;
+    const gp_Pnt p3 = (c.eFrame_eStringer.Location().XYZ() + c.sFrame_eStringer.Location().XYZ()) / 2;
+    const gp_Pnt p4 = (c.sFrame_eStringer.Location().XYZ() + c.sFrame_sStringer.Location().XYZ()) / 2;
+    const gp_Pnt p5 = (c.sFrame_sStringer.Location().XYZ() + c.eFrame_eStringer.Location().XYZ()) / 2;
+
+    Bnd_Box boundingBox;
+    BRepBndLib::Add(m_parent->GetParent()->GetParent()->GetParent()->GetLoft()->Shape(), boundingBox);
```

Comment:
What shape is put into to bounding box? Fuselage, Fuselage Structure?

Please annotate with a comment.

Reply:
I think @rmaierl can probably answer these questions, right?

Reply:
fuselage loft